File: algorithms/majority_element.rb
# @param {Integer[]} nums
# @return {Integer}
def majority_element(nums)
  count = 1
  m = nums[0]
  nums[1..-1].each do |x|
    if x == m
      count += 1
    else
      count -= 1
      if count == 0
        m = x
        count = 1
      end
    end
  end
  m
end

p majority_element([3, 2, 3])

File: algorithms/majority_element.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        count = 0
        ans = 0
        for x in nums:
            if count == 0:
                ans = x
            if ans == x:
                count += 1
            else:
                count -= 1
        return ans

t = Solution()

assert t.majorityElement([1, 1, 2]) == 1
assert t.majorityElement([1]) == 1

print("tests passed")
